services/spar/src/Spar/Data.hs

```diff
@@ -419,6 +429,47 @@ deleteIdPRawMetadata idp = retry x5 . write del $ params Quorum (Identity idp)
     del :: PrepQuery W (Identity SAML.IdPId) ()
     del = "DELETE FROM idp_raw_metadata WHERE id = ?"
 
+----------------------------------------------------------------------
+-- SSO settings
+
+-- It's important to maintain two invariants:
+-- 1) whenever there is a default code, it must also exist in the idp table
+-- 2) there can always only be one default SSO code selected
+
+getDefaultSSOCode
+  :: (HasCallStack, MonadClient m)
+  => m (Maybe SAML.IdPId)
+getDefaultSSOCode = fmap runIdentity . minimumMay <$>
+  (retry x1 . query sel $ params Quorum ())
+  where
+    sel :: PrepQuery R () (Identity SAML.IdPId)
+    sel = "SELECT idp FROM default_idp"
+
+storeDefaultSSOCode
+  :: (HasCallStack, MonadClient m)
+  => SAML.IdPId -> m ()
+storeDefaultSSOCode idpId = do
+  -- there is a race condition here which means there could potentially be more
+  -- than one entry (violating invariant 2).
+  -- However, the SELECT query will deterministally pick one of them and the
+  -- others will get removed by TRUNCATE the next time this function is called.
+  retry x5 . write trunc $ params Quorum ()
+  retry x5 . write ins $ params Quorum (Identity idpId)
```

Comment:
I couldn't put them into the same batch, as there is no ordering within a batch (and I didn't want to play around with `USING TIMESTAMP`). Would be fixed again by the other schema.